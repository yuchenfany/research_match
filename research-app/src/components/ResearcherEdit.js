/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import React from 'react';
import '../assets/index.css';
import { useNavigate } from 'react-router-dom';
import NavBar from './NavBar';

function ResearcherEdit({ user, setUser }) {
  const navigate = useNavigate();

  const updateResearcher = async (event) => {
    setUser({
      username: user.username,
      password: user.password,
      name: event.target.value,
      organization: user.organization,
      type: user.type,
    });
  };

  const updateOrganization = async (event) => {
    setUser({
      username: user.username,
      password: user.password,
      name: user.name,
      organization: event.target.value,
      type: user.type,
    });
  };

  async function postUserInfo() {
    await fetch(`http://localhost:5000/record/researcher-edit/${user.username}`, {
      method: 'POST',
      body: JSON.stringify(user),
      headers: {
        'Content-Type': 'application/json',
      },
    })
      .catch((e) => {
        window.alert(e);
      });

    return true;
  }

  async function handleSubmit(event) {
    if (await postUserInfo()) {
      navigate('/researcher-home');
    } else {
      event.preventDefault();
    }
  }

  async function handleCancel() {
    // retrieve stored state
    const stored = JSON.parse(localStorage.getItem(user.username));
    await setUser(stored);
    navigate('/researcher-home');
  }

  return (
    <div className="ResearcherEdit">
      <NavBar user={user} />
      <div className="profile-flex">
        <div className="header-left">Edit Researcher Profile</div>
        <div className="profile-row">
          <div className="input-label">Name</div>
          <input
            className="profile-input"
            type="text"
            id="researcher"
            value={user.name}
            onChange={updateResearcher}
          />
        </div>
        <div className="profile-row">
          <div className="input-label">Organization</div>
          <input
            className="profile-input"
            type="text"
            id="organization"
            value={user.organization}
            onChange={updateOrganization}
          />
        </div>
        <div className="profile-row">
          <div className="button-row">
            <input className="cancel-button" type="submit" value="CANCEL" onClick={handleCancel} />
            <input className="update-button" type="submit" value="UPDATE" onClick={handleSubmit} />
          </div>
        </div>
      </div>
    </div>
  );
}

export default ResearcherEdit;
