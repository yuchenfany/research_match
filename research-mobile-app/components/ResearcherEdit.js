/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import React, { useState } from 'react';
import {
  View, Button, Text, TextInput, StyleSheet,
} from 'react-native';
import NavBar from './NavBar';

function ResearcherEdit({ route, navigation }) {
  const { user, setUser } = route.params;

  const [tempUser, setTempUser] = useState({
    username: user.username,
    password: user.password,
    name: user.name,
    organization: user.organization,
    studies: user.studies,
    type: user.type,
    title: user.title,
  });

  // const [nameErr, setNameErr] = useState({ message: '' });
  // const [orgErr, setOrgErr] = useState({ message: '' });

  // const isValidInput = (input) => !(input.length === 0);

  const updateResearcher = async (event) => {
    await setTempUser({
      username: tempUser.username,
      password: tempUser.password,
      name: event.target.value,
      organization: tempUser.organization,
      studies: tempUser.studies,
      type: tempUser.type,
      title: tempUser.title,
    });
  };

  const updateOrganization = async (event) => {
    await setTempUser({
      username: tempUser.username,
      password: tempUser.password,
      name: tempUser.name,
      organization: event.target.value,
      studies: tempUser.studies,
      type: tempUser.type,
      title: tempUser.title,
    });
  };

  async function postUserInfo() {
    await fetch(`http://localhost:5000/record/researcher-edit/${user.username}`, {
      method: 'POST',
      body: JSON.stringify(tempUser),
      headers: {
        'Content-Type': 'application/json',
      },
    })
      .catch((e) => {
        throw new Error(e);
      });

    return true;
  }

  async function handleSubmit(event) {
    // if (!isValidInput(user.name) || !isValidInput(user.organization)) {
    //   if (!isValidInput(user.name)) {
    //     setNameErr({ message: 'Please enter your name' });
    //   } else {
    //     setNameErr({ message: '' });
    //   }

    //   if (!isValidInput(user.organization)) {
    //     setOrgErr({ message: 'Please enter your organization' });
    //   } else {
    //     setOrgErr({ message: '' });
    //   }

    //   event.preventDefault();

    //   return;
    // }

    if (await postUserInfo()) {
      await setUser(tempUser);

      navigation.navigate('ResearcherHome', {
        user: tempUser,
        setUser,
      });
    } else {
      event.preventDefault();
    }
  }

  // async function handleCancel() {
  //   // retrieve stored state
  //   const stored = JSON.parse(localStorage.getItem(user.username));
  //   await setUser(stored);
  //   // navigate('/researcher-home');
  // }

  const styles = StyleSheet.create({
    container: {
      backgroundColor: '#F3F8FA',
      flex: 1,
      padding: 20,
    },
    header: {
      fontSize: 20,
      lineHeight: 40,
      fontWeight: 500,
      marginTop: 30,
      color: '#103143',
    },
    subheader: {
      fontSize: 16,
      lineHeight: 40,
      fontWeight: 500,
      color: '#103143',
    },
    button: {
      width: 275,
      height: 35,
      fontSize: 12,
      letterSpacing: 1,
      marginTop: 20,
    },
    inputField: {
      width: 275,
      height: 27,
      backgroundColor: '#F9FFFE',
      borderColor: '#808A8F',
      borderWidth: 1,
      marginRight: 5,
      marginBottom: 10,
      borderRadius: 3,
      color: '#103143',
    },
  });

  return (
    <View style={styles.container}>
      <NavBar user={user} setUser={setUser} navigation={navigation} />
      <Text style={styles.header}>Edit Profile</Text>
      <Text color="#103143" marginTop="10">Name</Text>
      <TextInput
        style={styles.inputField}
        type="text"
        id="researcher"
        defaultValue={user.name}
        onChange={updateResearcher}
      />
      {/* <span className="error-message">{nameErr.message}</span> */}
      <Text color="#103143" marginTop="10">Organization</Text>
      <TextInput
        style={styles.inputField}
        type="text"
        id="organization"
        defaultValue={user.organization}
        onChange={updateOrganization}
      />
      {/* <span className="error-message">{orgErr.message}</span> */}
      {/* <input className="cancel-button" type="submit"
      value="CANCEL" onClick={handleCancel} /> */}
      <View style={styles.button}>
        <Button className="update-button" type="submit" title="UPDATE" color="#103143" onPress={(e) => handleSubmit(e)} />
      </View>
    </View>
  );
}

export default ResearcherEdit;
