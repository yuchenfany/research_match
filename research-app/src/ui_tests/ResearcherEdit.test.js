/**
 * @jest-environment jsdom
 */

/* eslint-disable react/prop-types */
/* eslint-disable react/jsx-filename-extension */
import { render, screen, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
// import '@testing-library/jest-dom';
import renderer from 'react-test-renderer';
import { BrowserRouter as Router } from 'react-router-dom';
import React from 'react';
// import { jest } from '@jest/globals';

import ResearcherEdit from '../components/ResearcherEdit';

// const mockedUsedNavigate = jest.fn();

// jest.mock('react-router-dom', () => ({
//   ...jest.requireActual('react-router-dom'),
//   useNavigate: () => mockedUsedNavigate,
// }));

test('Title Edit Researcher Profile', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  
  const notification = false;
  const mockSetStudy = jest.fn();
  const study = {
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  };
    render(
        <Router>
            <ResearcherEdit user={user1} setUser={mockSetUser} />
        </Router>);  
  const label = screen.getByText("Edit Researcher Profile");
  expect(label).toBeInTheDocument();
});

test('find name text', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  
  const notification = false;
  const mockSetStudy = jest.fn();
  const study = {
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  };
    render(
        <Router>
            <ResearcherEdit user={user1} setUser={mockSetUser} />
        </Router>);  
  const label = screen.getByText("Name");
  expect(label).toBeInTheDocument();
});

test('find organization text', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  
  const notification = false;
  const mockSetStudy = jest.fn();
  const study = {
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  };
    render(
        <Router>
            <ResearcherEdit user={user1} setUser={mockSetUser} />
        </Router>);  
  const label = screen.getByText("Organization");
  expect(label).toBeInTheDocument();
});

test('single input value  sequence', () => {
  const mockUser = {
    username: '', password: '',
  };
  const mockSetUser = jest.fn();

  render(
    <Router>
      <ResearcherEdit user={mockUser} setUser={mockSetUser} />
    </Router>,
  );

  const input = screen.getByPlaceholderText('name');
  fireEvent.change(input, { target: { value: 'user1' } });
  expect(input.value).toEqual('user1');
  expect(mockSetUser).toHaveBeenCalled();

  const input2 = screen.getByPlaceholderText('organization');
  fireEvent.change(input2, { target: { value: 'user1pass' } });
  expect(input2.value).toEqual('user1pass');
  expect(mockSetUser).toHaveBeenCalled();

  // fireEvent.click(screen.getByDisplayValue('SUBMIT'));
});

test('Update Button good name',() => {
  const Button = ({onClick, children}) => (
    <button onClick={onClick}>{children}</button>
  )
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass', name: 'hiya', organization: "UPenn" };
    render(
        <Router>
            <ResearcherEdit user={user1} setUser={mockSetUser} />
        </Router>);  
  const linkElement = screen.getByRole('button', {
    name: /update/i,
  });
  fireEvent.click(linkElement);
  expect(linkElement).toBeInTheDocument();
});

test('Update Button empty name',() => {
  const Button = ({onClick, children}) => (
    <button onClick={onClick}>{children}</button>
  )
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass', name: '', organization: 'UPenn' };
    render(
        <Router>
            <ResearcherEdit user={user1} setUser={mockSetUser} />
        </Router>);  
  const linkElement = screen.getByRole('button', {
    name: /update/i,
  });
  fireEvent.click(linkElement);
  expect(linkElement).toBeInTheDocument();
});

test('Update Button empty organization',() => {
  const Button = ({onClick, children}) => (
    <button onClick={onClick}>{children}</button>
  )
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass', name: 'yes sirr', organization: '' };
    render(
        <Router>
            <ResearcherEdit user={user1} setUser={mockSetUser} />
        </Router>);  
  const linkElement = screen.getByRole('button', {
    name: /update/i,
  });
  fireEvent.click(linkElement);
  expect(linkElement).toBeInTheDocument();
});

test('Cancel Button',() => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
    render(
        <Router>
            <ResearcherEdit user={user1} setUser={mockSetUser} />
        </Router>);  
  const linkElement = screen.getByRole('button', {
    name: /cancel/i,
  });
  fireEvent.click(linkElement);
  expect(linkElement).toBeInTheDocument();
});


test('snapshot test', () => {
  const mockSetUser = jest.fn();
  const user1 = { username: 'user1', password: 'user1pass' };
  
  const notification = false;
  const mockSetStudy = jest.fn();
  const study = {
    studyId: '',
    participants: [''],
    researchers: [],
    tags: [],
  };
  const component = renderer.create( 
    <Router>
        <ResearcherEdit user={user1} setUser={mockSetUser} />
    </Router>);
  const tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
//TO GET HIGHER SCORE

//1. fireevent update button can not be clicked
