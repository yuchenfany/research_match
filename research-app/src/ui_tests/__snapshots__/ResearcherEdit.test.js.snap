// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot test 1`] = `
<div
  className="ResearcherEdit"
>
  <div
    className="nav"
  >
    <button
      className="nav-btn"
      onClick={[Function]}
      type="button"
    >
      HOME
    </button>
    <button
      className="nav-btn"
      onClick={[Function]}
      type="button"
    >
      MY STUDIES
    </button>
    <button
      className="nav-btn"
      onClick={[Function]}
      type="button"
    >
      MESSAGES
    </button>
    <button
      className="nav-btn"
      onClick={[Function]}
      type="button"
    >
      ANALYTICS
    </button>
    <button
      className="account-btn"
      onClick={[Function]}
      type="button"
    >
      ACCOUNT
    </button>
    <div
      className="acct-options"
      style={
        Object {
          "display": "none",
        }
      }
    >
      <button
        className="account-btn"
        onClick={[Function]}
        type="button"
      >
        EDIT PROFILE
      </button>
      <button
        className="account-btn"
        onClick={[Function]}
        type="button"
      >
        DELETE ACCOUNT
      </button>
      <button
        className="logout-btn"
        onClick={[Function]}
        type="button"
      >
        LOGOUT
      </button>
    </div>
  </div>
  <div
    className="profile-flex"
  >
    <div
      className="header-left"
    >
      Edit Researcher Profile
    </div>
    <div
      className="profile-row"
    >
      <div
        className="input-label"
      >
        Name
      </div>
      <input
        className="profile-input"
        id="researcher"
        onChange={[Function]}
        placeholder="name"
        type="text"
      />
    </div>
    <div
      className="profile-row"
    >
      <span
        className="error-message"
      >
        
      </span>
    </div>
    <div
      className="profile-row"
    >
      <div
        className="input-label"
      >
        Organization
      </div>
      <input
        className="profile-input"
        id="organization"
        onChange={[Function]}
        placeholder="organization"
        type="text"
      />
    </div>
    <div
      className="profile-row"
    >
      <span
        className="error-message"
      >
        
      </span>
    </div>
    <div
      className="profile-row"
    >
      <div
        className="button-row"
      >
        <button
          className="cancel-button"
          onClick={[Function]}
          type="submit"
          value="CANCEL"
        >
           CANCEL
        </button>
        <button
          className="update-button"
          onClick={[Function]}
          type="submit"
          value="UPDATE"
        >
           UPDATE
        </button>
      </div>
    </div>
  </div>
</div>
`;
